File: ruby/test/uptimeTest.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../src/uptime'

class UptimeTest < Minitest::Test
  def setup
    @uptime_utility = Uptime.new
  end

  def test_convert_sec
    lambda {
      sec = 777_600 # 9 days
      assert_equal @uptime_utility.convert_sec(sec), [9, 0, 0]
    }.call

    lambda {
      sec = 54_000 # 15 hours
      assert_equal @uptime_utility.convert_sec(sec), [0, 15, 0]
    }.call

    lambda {
      sec = 3599 # 59 minutes
      assert_equal(@uptime_utility.convert_sec(sec), [0, 0, 59])
    }.call

  end

  def test_format_day
    lambda {
      number_of_days = 0
      assert_equal @uptime_utility.format_day(number_of_days), ''
    }.call

    lambda {
      number_of_days = 1
      assert_equal @uptime_utility.format_day(number_of_days), '1 day, '
    }.call

    lambda {
      number_of_days = 20
      assert_equal @uptime_utility.format_day(number_of_days), '20 days, '
    }.call
  end

  def test_format_hours_and_minutes
    lambda {
      number_of_hour = 0
      number_of_minute = 0
      assert_equal @uptime_utility.format_hours_and_minutes(number_of_hour, number_of_minute), '0 min'
    }.call

    lambda {
      number_of_hour = 7
      number_of_minute = 5
      assert_equal @uptime_utility.format_hours_and_minutes(number_of_hour, number_of_minute), '7:05'
    }.call

    lambda {
      number_of_hour = 17
      number_of_minute = 20
      assert_equal @uptime_utility.format_hours_and_minutes(number_of_hour, number_of_minute), '17:20'
    }.call
  end

  def test_generate_file_content
    lambda {
      up_sec = 86461 # 1 day 00:01:01.
      idle_sec = 23456789 # 271 days 11:46:29.
      assert_match('1 day, 1 min', @uptime_utility.generate_file_content(up_sec, idle_sec))
      assert_match('271 days, 11:46', @uptime_utility.generate_file_content(up_sec, idle_sec))
    }.call

    lambda {
      up_sec = 4567 # 01:16:07.
      idle_sec = 4056 # 01:07:36.
      assert_match('1:16', @uptime_utility.generate_file_content(up_sec, idle_sec))
      assert_match('1:07', @uptime_utility.generate_file_content(up_sec, idle_sec))
    }.call
  end
end

File: cpp/test/uptimeTest.cpp
//
// Created by theow on 26/10/2021.
//

#include "../include/uptime.hpp"
#include "doctest.h"
#include <unistd.h>
#include <regex>

TEST_CASE ("convert_sec()") {
    {
        int array[3];
        int sec = 777600;
        uptime::convert_sec(sec, array);
        int expected[3] = {9, 0, 0};
        bool result = (array[0] == expected[0] && array[1] == expected[1] && array[2] == expected[2]);
                CHECK(result);
    }
    {
        int array[3];
        int sec = 7200;
        uptime::convert_sec(sec, array);
        int expected[3] = {0, 2, 0};
        bool result = (array[0] == expected[0] && array[1] == expected[1] && array[2] == expected[2]);
                CHECK(result);
    }
    {
        int array[3];
        int sec = 120;
        uptime::convert_sec(sec, array);
        int expected[3] = {0, 0, 2};
        bool result = (array[0] == expected[0] && array[1] == expected[1] && array[2] == expected[2]);
                CHECK(result);
    }
}


TEST_CASE ("format_day") {
    {
        int number_of_day = 0;
        std::string result = uptime::format_day(number_of_day);
                CHECK(result == "");
    }
    {
        int number_of_day = 1;
        std::string result = uptime::format_day(number_of_day);
                CHECK(result == "1 day, ");
    }
    {
        int number_of_day = 20;
        std::string result = uptime::format_day(number_of_day);
                CHECK(result == "20 days, ");
    }
}

TEST_CASE ("format_hours_and_minutes") {
    {
        int number_of_hour = 0;
        int number_of_minute = 0;
        std::string result = uptime::format_hours_and_minutes(number_of_hour, number_of_minute);
                CHECK(result == "0 min");
    }
    {
        int number_of_hour = 7;
        int number_of_minute = 5;
        std::string result = uptime::format_hours_and_minutes(number_of_hour, number_of_minute);
                CHECK(result == "7:05");
    }
    {
        int number_of_hour = 17;
        int number_of_minute = 20;
        std::string result = uptime::format_hours_and_minutes(number_of_hour, number_of_minute);
                CHECK(result == "17:20");
    }
}
/*
 * Way too hard to implement without a proper regex support, yes, i'm talking about you c++ 11...
TEST_CASE ("generate_file_content") {
    {

        int up_sec = 86461;
        int idle_sec = 23456789;
        std::string result = uptime::generate_file_content(up_sec, idle_sec);
        CHECK(result == " up 1 day, 1 min idle 271 days, 11:46");
    }

    {
        int up_sec = 4567;
        int idle_sec = 4056;
        std::string result = uptime::generate_file_content(up_sec, idle_sec);
        std::cout << result;
        std::regex e("[0-9]{2}:[0-9]{2}:[0-9]{2} up 1:16\n[0-9]{2}:[0-9]{2}:[0-9]{2} idle 1:07");
        bool value = std::regex_match(result, e);
                CHECK(value);
    }
}
 */